libs/ngx-nuts-and-bolts/src/lib/utilities/table-state/table-state.spec.ts

```diff
@@ -0,0 +1,195 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
+/* eslint-disable */
+/* eslint-disable */
+import { TestBed } from '@angular/core/testing';
+import { ActivatedRoute, Router } from '@angular/router';
+import { RouterTestingModule } from '@angular/router/testing';
+import { take } from 'rxjs/operators';
+import {
+	changeFilters,
+	changePage,
+	changeSort,
+	createCustomFiltersObservable,
+	createPaginationObservable,
+	createSortObservable,
+	IPageInfo,
+	ISortInfo,
+} from './table-state';
+
+describe('Table sate', () => {
+	let navigationSpy: jest.SpyInstance<Promise<boolean>>;
+	let router: Router;
+	let route: ActivatedRoute;
+
+	beforeEach(async () => {
+		await TestBed.configureTestingModule({
+			imports: [RouterTestingModule],
+		});
+
+		router = TestBed.inject(Router);
+		route = TestBed.inject(ActivatedRoute);
+
+		navigationSpy = jest.spyOn(router, 'navigate');
+	});
+
+	it('should set query parameters for pagination', async () => {
+		expect(navigationSpy).toHaveBeenCalledTimes(0);
+		let paginationValue = await createPaginationObservable(route).pipe(take(1)).toPromise();
+
+		const pageInfo: IPageInfo = {
+			pageIndex: 1,
+			pageSize: 12,
+		};
+
+		await changePage(router, pageInfo);
+		paginationValue = await createPaginationObservable(route).pipe(take(1)).toPromise();
+
+		expect(navigationSpy).toHaveBeenCalledTimes(1);
+		expect(paginationValue).not.toBe(null);
+		expect(paginationValue).toEqual(pageInfo);
+	});
+
+	it('should return sorting info after page sort info change', async () => {
+		expect(navigationSpy).toHaveBeenCalledTimes(0);
+		let sortValue = await createSortObservable(route).pipe(take(1)).toPromise();
+
+		const sortInfo: ISortInfo = {
+			sortDirection: 'asc',
+			sortKey: 'testing-key',
+		};
+
+		await changeSort(router, sortInfo);
+
+		sortValue = await createSortObservable(route).pipe(take(1)).toPromise();
+
+		expect(navigationSpy).toHaveBeenCalledTimes(1);
+		expect(sortValue).not.toBe(null);
+		expect(sortValue).toEqual(sortInfo);
+	});
+
+	it('should return filters', async () => {
+		let filtersValue = await createCustomFiltersObservable(route, router).pipe(take(1)).toPromise();
+		expect(navigationSpy).toHaveBeenCalledTimes(0);
+		const filter = {
+			arrayQuery: ['foo', 'bar'],
+			textQuery: 'baz',
+			numberQuery: 42,
+			objectQuery: { xyz: 'ijk' },
+		};
+
+		expect(filtersValue).toBe(null);
+
+		await changeFilters(router, filter);
+
+		filtersValue = await createCustomFiltersObservable(route, router).pipe(take(1)).toPromise();
```

Comment:
I don't thinks that creating a new observable is the best way to test this. What if there is a bug where the observable does not get a new value if it was created before changePage is called? This test would not catch that bug.

I think something like this would be better (and you might not even need await, depending on how the mock router behaves):

```ts
const callbacks = { next: jest.fn(), error: jest.fn(), complete: jest.fn() };
expect(callbacks.next).toHaveBeenCalledTimes(0);

createPaginationObservable(route).subscribe(callbacks);

expect(callbacks.next).toHaveBeenCalledTimes(1);
expect(callbacks.next.calls[0].args).toBe(null); // or whatever the initial value is;

const pageInfo: IPageInfo = {...};
await changePage(...);

expect(callbacks.next).toHaveBeenCalledTimes(2);
expect(callbacks.next.calls[1].args).toEqual(pageInfo);
expect(callbacks.error).toHaveBeenCalledTimes(0);
expect(callbacks.complete).toHaveBeenCalledTimes(0);
```

Please do the same for other tests as well.